### PR TITLE
Issue #7754: Update AbstractChecks to log DetailAST - OneTopLevelClass

### LINKED ISSUE #7754: Update AbstractChecks to log DetailAST - OneTopLevelClass
Child of https://github.com/checkstyle/checkstyle/issues/5777

This issue specifically focuses on OneTopLevelClass.

### PR DESCRIPTION
**Pull request by wltan**
Resolve #7754: Update AbstractChecks to log DetailAST - OneTopLevelClass

Diff report: https://wltan.github.io/checkstyle-reports/2020-03-08/onetoplevel/diff/

There is only one log call to fix: https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java#L153

Notes:
1. I wasn't able to upload all of the "deeper" files due to the max file path length (on Windows). But I think all of the essential information should be there.
2. All of the discrepancies seem to be about the extra column number, which is expected as we now log the DetailAST node instead of just the line number.
3. Our unit tests only cover violations at column 1, but other projects can have violations at other columns as well (e.g. `apache-ant` has a violation at column 2 instead as there is a whitespace in front of the declaration for whatever reason). I think we should update our tests to reflect this?

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java [hunk 1]
```diff
@@ -98,8 +99,9 @@
      */
     private boolean publicTypeFound;
 
-    /** Mapping between type names and line numbers of the type declarations.*/
-    private final SortedMap<Integer, String> lineNumberTypeMap = new TreeMap<>();
+    /** Mapping between type names and DetailAST nodes of the type declarations.*/
+    private final SortedMap<DetailAST, String> lineNumberTypeMap =
+            new TreeMap<>(Comparator.comparingInt(DetailAST::getLineNo));
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java]: I doubt it should be like this. In very rare case, e.g.
`enum A {}; enum B{}` no violation will be reported. 
Please change it.

Comment (wltan) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java]: I changed the comparator to also check the column number if the line number is equal. That should be good enough to ensure the nodes are unique and still in order.

**Comment by wltan** [comment]
Diff report has been updated in the first comment.
Old: https://wltan.github.io/checkstyle-reports/2020-03-07/onetoplevel/diff/
New: https://wltan.github.io/checkstyle-reports/2020-03-08/onetoplevel/diff/

There is one extra violation coming from the changes requested above. As expected, it is caused by two top-level types on the same line: https://wltan.github.io/checkstyle-reports/2020-03-08/onetoplevel/diff/lombok-ast/index.html#A143

**Comment by rnveach** [comment]
> There is one extra violation 

Good find. `Number of unique base messages reported below` and `Number of unique path messages reported below` should match for each project otherwise it is saying there is an uneven amount of changes, either new violations or violations removed. We aren't expecting this type of change unless you can account for it. I will update main issue with this note.

**Comment by rnveach** [comment]
> I wasn't able to upload all of the "deeper" files due to the max file path length (on Windows). But I think all of the essential information should be there.

See https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#command-line-arguments
> shortFilePaths (s) - whether to save report file paths as a shorter version to prevent long paths. This option is useful for Windows users where they are restricted to maximum directory depth (optional, default is false).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java [hunk 2]
```diff
@@ -92,14 +93,19 @@
      */
     public static final String MSG_KEY = "one.top.level.class";
 
+    /** Compare DetailAST nodes by line and then column number to make a unique ordering. */
+    private static final Comparator<DetailAST> LINE_AND_COL_COMPARATOR =
+            Comparator.comparingInt(DetailAST::getLineNo).thenComparingInt(DetailAST::getColumnNo);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java]: I am lost where the unique ordering is needed. Checkstyle will drop violations for the same line/column/message/check. It also already sorts violations on line/column. Can you explain what I am missing?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheckTest.java [hunk 3]
```diff
@@ -132,43 +132,66 @@ public void testFileWithThreeTopLevelEnum() throws Exception {
...
+            "11:1: " + getCheckMessage(MSG_KEY, "InputOneTopLevelClassEnum2inner2"),
         };
         verify(checkConfig, getPath("InputOneTopLevelClassEnum2.java"), expected);
     }
 
     @Test
     public void testFileWithFewTopLevelClasses() throws Exception {
         final DefaultConfiguration checkConfig =
-            createModuleConfig(OneTopLevelClassCheck.class);
+                createModuleConfig(OneTopLevelClassCheck.class);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheckTest.java]: Undo unnecessary changes like this. I see others in the same diff.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java (hunk 2 above)
Comment (wltan) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java]: The sorting is not so much for the order of the violations, but for this rule:
> If file doesn't contains public class, enum or interface, top-level type is the first type in file.

DetailAST on its own is not comparable, so we have to apply our own comparator to determine which is the first type in the file. This comes from extending the existing TreeMap to remember not just the line number, but the whole DetailAST node as well.

In hindsight, traversing the AST in `beginTree` should already visit the nodes in order. Maybe we can use a more efficient data structure instead? I'm thinking of using some sort of List; I just need a `List<DetailAST>` that stores the nodes in the traversed (sorted) order and I can get the individual type name from the node itself later.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheckTest.java (hunk 3 above)
Comment (wltan) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheckTest.java]: Fixed (I think I got all of them)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java (hunk 2 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java]: Thanks. I see now. Original map sorted on integer which is the key.

> Maybe we can use a more efficient data structure instead

Feel free to create a new issue and PR for this change.

**Comment by rnveach** [comment]
Can be merged when CI passes.

**Comment by rnveach** [comment]
Please rebase on latest master to resolve CI spellchecker issue.
